### PR TITLE
new endpoint for frozen-incidence

### LINKED ISSUE #148: History incidence not correct
The history data of the weekly incidence is not correct:
image

Source: https://api.corona-zahlen.org/districts/09172/history/incidence/3

But if i compare it with the data of the district office, there is some differences.: 

image

Greetings

### LINKED ISSUE #204: frozen incidence history
rki publishes the frozen incidence values in an excel file. This file is also used for the "emergency brake" (Notbremse).  Please look here [rki.de/inzidenzen](https://rki.de/inzidenzen).
i have written a new file in the data-requests folder. this change would fix #148 
Please take a look: [incidence.ts](https://github.com/PatrickHaussmann/rki-covid-api/blob/7a602774781237c1991bac3a83fbe3493fa0d0c8/src/data-requests/incidence.ts)

i am very happy to write a pull request, i just need help to structure the changes and split them into the necessary files and functions.

I would also find it very helpful to return the status of the emergency brake for each district as suggested in the linked file, as it is relatively complicated to calculate.

### PR DESCRIPTION
**Pull request by PatrickHaussmann**
the discussion for this pull request is in #204. also, it would fix problems like in #148 

rki publishes the frozen incidence values in an excel file. This file is also used for the "emergency brake" (Notbremse). 
Please look here [rki.de/inzidenzen](https://rki.de/inzidenzen).

Todos before it can be merged
- [x] write functions for `data-requests`
- [x] write functions for `responses`
- [x] move function from `src/responses/frozen-incidence.ts` to `src/responses/districts.ts`
- [x] document changes in `/docs/endpoints/districts.md` and `/README.md`
- [x] implement `/districts/history/frozen-incidence/:days`
- [x] implement `/districts/:district/history/frozen-incidence`
- [x] implement `/districts/:district/history/frozen-incidence/:days`
- [ ] document the difference to e.g. `/districts/history/incidence`
- [ ] direct users to use this endpoint if they want data used for political decisions or in TV/newspapers

fix #148 
fix #204

### CODE REVIEW

**Comment by PatrickHaussmann** [comment]
after looking over it again, i thought that `src/responses/frozen-incidence.ts` should probably not be an extra file, rather it should be added to `src/responses/districts.ts`

**Comment by Rubber1Duck** [comment]
> after looking over it again, i thought that `src/responses/frozen-incidence.ts` should probably not be an extra file, rather it should be added to `src/responses/districts.ts`

good point, make sense

>     * [ ]  implement `/districts/history/:days/incidence`

do you meen `/districts/history/frozen-incidence/:days` (argument :days is after incidence!)
`/districts/history/incidence/:days` already exists!

then maybe `/districts/:district/history/frozen-incidence` and `/districts/:district/history/frozen-incidence/:days` are missing, too

you have to edit `/docs/.vuepress/config.js` and README.md, too.

**Comment by Rubber1Duck** [comment]
> you have to edit `/docs/.vuepress/config.js` and README.md, too.

sorry, if you move the functions to district.ts edit of `/docs/.vuepress/config.js` is not nessesary! because the new  links are part of /districts

**Comment by Rubber1Duck** [comment]
> sorry, if you move the functions to district.ts edit of `/docs/.vuepress/config.js` is not nessesary! because the new links are part of /districts

For README.md also

**Review by Rubber1Duck — APPROVED** [review]
so far, looks fine

**Comment by PatrickHaussmann** [comment]
The code changes are done so far (except for appreciated suggested changes).
What are your comments?

I think i need help documenting and explaining what the difference is to `incidence`. 
I'm also not sure in which places this should best be done

File: README.md [hunk 1]
```diff
@@ -71,6 +71,10 @@ https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/Projekte_RKI/Nowcas
 
 https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/Daten/Testzahlen-gesamt.xlsx?__blob=publicationFile
 
+**Incidence data**
```
Comment (Rubber1Duck) [inline README.md]: That should be `**Frozen-Incidence data**`
the other updating incidences are calculated from weekly new cases

**Review by Rubber1Duck — CHANGES_REQUESTED** [review]

**Comment by Rubber1Duck** [comment]
> I think i need help documenting and explaining what the difference is to `incidence`.
> I'm also not sure in which places this should best be done

here is the german text witch described the difference very well, "just" translate it :-)
Gemäß § 28b IfSG veröffentlicht das RKI für alle Landkreise und kreisfreien Städte fortlaufend die 7-Tage-Inzidenz.
Bei den 7-Tage-Inzidenzen für frühere Tage muss berücksichtigt werden, dass es sich um die jeweils an dem angegebenen Tag berichteten Werte handelt, die nicht durch an Folgetagen nachübermittelte Fälle aktualisiert werden (für den Berichtstag "eingefrorene" Werte).
Diese Werte sind für das In- und Außerkrafttreten der bundeseinheitlichen Maßnahmen nach § 28b des Infektionsschutzgesetzes („Bundesnotbremse“) maßgeblich. Die Zugrundelegung der „eingefrorenen“ Werte stellt sicher, dass die Werte keinen Schwankungen unterliegen und sich die von den Maßnahmen Betroffenen auf das In- bzw. Außerkrafttreten dieser mit einem zeitlichen Vorlauf einstellen können.
Da in der Regel an den Folgetagen noch Fälle mit einem Meldedatum an einem der Vortage nachübermittelt werden, erhöht sich die 7-Tage-Inzidenz bei rückwirkender Betrachtung. Diese Inzidenzwerte entsprechen den "alten" Inzidenzwerten! welche unter den Teillinks xxxxxxxxx/incidence/xxxxxxxx verfügbar sind.

this should be part of frozen-incidence.md i think !?

File: src/data-requests/frozen-incidence.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  const lastUpdate = new Date(dateString);
+
+  if (ags != null) {
+    // if ags is not defined restrict days to 36
+    if (days != null) {
+      days = Math.min(days, 36);
+    } else {
+      days = 36;
+    }
+  }
```
Comment (marlon360) [inline src/data-requests/frozen-incidence.ts]: We don't need to limit the days (this is only needed for the arcgis endpoints)

File: src/data-requests/frozen-incidence.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    }
+  );
+  const data = response.data;
+  if (data.error) {
+    throw new RKIError(data.error, response.config.url);
+  }
+
+  var workbook = XLSX.read(data, { type: "buffer", cellDates: true });
+  const sheet = workbook.Sheets["LK_7-Tage-Inzidenz"];
+  const json = XLSX.utils.sheet_to_json(sheet);
```
Comment (marlon360) [inline src/data-requests/frozen-incidence.ts]: We should parse the sheet in a better way. The keys of the object should be the headers.

File: src/data-requests/frozen-incidence.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    }
+  }
+
+  const districts = json.map((element) => {
+    // one district in each row
+
+    const keys = Object.keys(element); // columns
+    keys.shift(); // rowNumber
+    const name = element[keys.shift()];
+    const ags = element[keys.shift()].toString().padStart(5, "0");
```
Comment (marlon360) [inline src/data-requests/frozen-incidence.ts]: Working with shift unpredictable, we should access the row by name.

**Review by marlon360 — CHANGES_REQUESTED** [review]
I would like to change the way we parse the file a bit. I am going to prepare something.

**Comment by marlon360** [comment]
```TS
import axios from "axios";
import XLSX from "xlsx";
import { getDateBefore, RKIError } from "../utils";
import { ResponseData } from "./response-data";

export interface FrozenIncidenceData {
  ags: string;
  name: string;
  history: {
    date: string;
    value: number;
  }[];
}

export async function getFrozenIncidenceHistory(
  days?: number,
  ags?: string
): Promise<ResponseData<FrozenIncidenceData[]>> {
  const response = await axios.get(
    "https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/Daten/Fallzahlen_Kum_Tab.xlsx?__blob=publicationFile",
    {
      responseType: "arraybuffer",
    }
  );
  const data = response.data;
  if (data.error) {
    throw new RKIError(data.error, response.config.url);
  }

  var workbook = XLSX.read(data, { type: "buffer", cellDates: true });
  const sheet = workbook.Sheets["LK_7-Tage-Inzidenz"];
  // table starts in row 5 (parameter is zero indexed)
  const json = XLSX.utils.sheet_to_json(sheet, { range: 4 });

  // date is in cell A2
  const dateString = sheet["A2"].v.replace("Stand: ", "");
  const lastUpdate = new Date(dateString);

  const districts = json.map((district) => {
    const name = district["LK"];
    const ags = district["LKNR"].toString().padStart(5, "0");

    let history = [];

    // get all date keys
    const dateKeys = Object.keys(district);
    // ignore the first three elements (rowNumber, LK, LKNR)
    dateKeys.splice(0, 3);
    dateKeys.forEach((dateKey) => {
      const date_pattern = /(\d{2})\.(\d{2})\.(\d{4})/;
      const date = new Date(
        dateKey.toString().replace(date_pattern, "$3-$2-$1")
      );
      history.push({ weekIncidence: district[dateKey], date });
    });

    if (days != null) {
      const reference_date = new Date(getDateBefore(days));
      history = history.filter((element) => element.date > reference_date);
    }

    return { ags, name, history };
  });

  return {
    data: districts,
    lastUpdate: lastUpdate,
  };
}

```

**Comment by PatrickHaussmann** [comment]
> I would like to change the way we parse the file a bit. I am going to prepare something.

thank you for these very good improvement suggestions

File: src/responses/districts.ts [hunk 5]
```diff
@@ -372,3 +376,31 @@ export async function DistrictsRecoveredHistoryResponse(
...
+
+  const data: {
+    [key: string]: FrozenIncidenceData;
+  } = {};
+
+  for (const element of FrozenIncidenceData.data) {
+    if (ags && element.ags != ags) continue;
+
+    data[element.ags] = element;
+  }
```
Comment (marlon360) [inline src/responses/districts.ts]: Let's move the `ags` filtering to the data-request and use `filter` of `Array`.

File: src/data-requests/frozen-incidence.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    });
+
+    if (days != null) {
+      const reference_date = new Date(getDateBefore(days));
+      history = history.filter((element) => element.date > reference_date);
+    }
+
+    return { ags, name, history };
+  });
+
```
Comment (marlon360) [inline src/data-requests/frozen-incidence.ts:64]: Let's add the `ags` filtering here.

**Review by marlon360 — CHANGES_REQUESTED** [review]
I'd like to move the `ags` filtering to `getFrozenIncidenceHistory`.

**Comment by marlon360** [comment]
districts.ts:
```TS
interface FrozenIncidenceHistoryData extends IResponseMeta {
  data: {
    [key: string]: FrozenIncidenceData
  }
}

export async function FrozenIncidenceHistoryResponse(
  days?: number,
  ags?: string
): Promise<FrozenIncidenceHistoryData> {
  const frozenIncidenceHistoryData = await getFrozenIncidenceHistory(days, ags);

  let data = {};
  frozenIncidenceHistoryData.data.forEach(historyData => {
    data[historyData.ags] = historyData
  });

  return {
    data: data,
    meta: new ResponseMeta(frozenIncidenceHistoryData.lastUpdate),
  };
}

```

frozen-incidence.ts:

```TS
  let districts = json.map((district) => {
    const name = district["LK"];
    const ags = district["LKNR"].toString().padStart(5, "0");

    let history = [];

    // get all date keys
    const dateKeys = Object.keys(district);
    // ignore the first three elements (rowNumber, LK, LKNR)
    dateKeys.splice(0, 3);
    dateKeys.forEach((dateKey) => {
      const date_pattern = /(\d{2})\.(\d{2})\.(\d{4})/;
      const date = new Date(
        dateKey.toString().replace(date_pattern, "$3-$2-$1")
      );
      history.push({ weekIncidence: district[dateKey], date });
    });

    if (days != null) {
      const reference_date = new Date(getDateBefore(days));
      history = history.filter((element) => element.date > reference_date);
    }

    return { ags, name, history };
  });


  if (ags != null) {
    districts = districts.filter((district) => district.ags === ags);
  }
```

**Review by marlon360 — APPROVED** [review]

**Comment by marlon360** [comment]
Great job @PatrickHaussmann!
We can add the descriptions in the docs later.

**Comment by Rubber1Duck** [comment]
unfortnatly the list is unsorted!
frozen

**Comment by marlon360** [comment]
The list is sorted alphabetically by name.

**Comment by Rubber1Duck** [comment]
sure?
frozen2
but if so thats not in line with all the other lists!

**Comment by Rubber1Duck** [comment]
is it a big problem to resort this by the key?

**Comment by PatrickHaussmann** [comment]
I believe that this is not possible. 
the order of an object "does not depend on how an object is defined" [source](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries).
[or here is another good discussion about this](https://stackoverflow.com/questions/30076219/does-es6-introduce-a-well-defined-order-of-enumeration-for-object-properties)

another possibility I could imagine is write a custom `toJSON` function to be used by `res.json` in `server.ts` in which we could sort the entries by ags


it may well be that I am overlooking something

**Comment by marlon360** [comment]
I think it does not make sense to order keys. Only arrays should have a defined order (e.g. the history array).
We use keys to access certain districts directly, without having to loop through them.

If you want to order the keys, you can do it inside your application.
- get key array
- order keys the way you want
- loop through ordered keys and access associated value

**Comment by Rubber1Duck** [comment]
Okay

**Pull request merged**